### PR TITLE
Task objects with Annotations

### PR DESCRIPTION
**Pull request by sjperkins**
- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

### CODE REVIEW

**Comment by sjperkins** [comment]
Adds a task object slotted on ``function``, ``args``, ``kwargs`` and ``annotations``.

Notes regarding this implementation:

1. ``Task.from_spec`` converts task tuples into Task objects, and recursively applies this conversion on nested python types such as lists and dicts, according to https://docs.dask.org/en/latest/spec.html#definitions.
2. ``dsk = Task.from_spec(dsk)`` lines are hard-coded into dask.local.get_async and dask.core.get. This exists to prove that ``Task.from_spec`` successfuly translates dask graphs in the test suite. They should be removed if this PR progresses.
3. Old style tasks and new style task should be able to intermingle.
4. Benchmarking is needed, obviously once the lines in (2) have been removed.

**Comment by jakirkham** [comment]
cc @madsbk (who may find this of interest ;)

**Comment by sjperkins** [comment]
I'm pretty happy with how this has turned out, especially considering that the test cases pass with a simple statement converting a graph containing tuple tasks into a graph containing the new Task objects. Therefore, I'm confident that the internals of the local schedulers handle the new Task objects correctly.

However, this avoids passing new Task objects through the optimization routines, so the next bit of work on this PR will need to focus on:

1. optimize_blockwise
2. fuse_linear
3. fuse
 
I'll probably kick that off by writing a ``Task.subs`` or ``Task.fuse`` method, towards the end of this week.

I think I will find (3) challenging. Is there anyone who knows this part of the code well and who may be able to advise or help?

/cc @rjzamora, @mrocklin, @eriknw, @dhirschfeld, @TomAugspurger @mrocklin @o-smirnov

**Comment by eriknw** [comment]
@sjperkins I know `fuse` and `fuse_linear` very well and can help as you run into problems.  If you can get the dependencies of tasks and perform `subs`, then I would expect the rest to more or less just work.

**Comment by sjperkins** [comment]
> 
> 
> @sjperkins I know `fuse` and `fuse_linear` very well and can help as you run into problems. If you can get the dependencies of tasks and perform `subs`, then I would expect the rest to more or less just work.

Ah thanks for that.

I feel confident about finding dependencies and writing a subs  so this might be easier than I expected.

Will shout if I run into anything tricky.

**Comment by mrocklin** [comment]
Thank you as always for being game to try these things out @sjperkins .  

I'm cc'ing @kkraus14 , @quasiben , and @kobzol here who have an interest in this topic.  To set expectations, there are enough people engaged here that I don't expect to converge on a solution very quickly.  However, having someone drive forward an approach is, I think, critical to getting people focused on raising concerns and requirements that they may have.  Thank you for pushing on this @sjperkins .

**Comment by jakirkham** [comment]
AIUI that's what Mads is trying to do in issue ( https://github.com/dask/distributed/issues/3811 ). Though please correct me if I'm misunderstanding Mads :)

**Comment by sjperkins** [comment]
@mrocklin Sure, I'm happy to drive this forward.

From my side, my goals are to:

- [x] 1. Create Task objects supporting Annotations 
- [x] 2. Support Tasks within local scheduler code (without respecting annotations)
- [ ] 3. Support Tasks within optimization code.
- [ ] 4. Support Task's within distributed scheduler code (respecting annotations resembling existing Client.submit kwargs such as workers, priority, resources etc.). Mostly done in https://github.com/dask/distributed/pull/3796

with the intention of writing Scheduler Plugins with update_graph methods that perform custom scheduling based on Annotations.

The above may seem like extensive changes to support custom scheduling code. However, the structure I've suggested intentionally avoids extensively modifying the internal local + distributed scheduling code because:

1. Modifying existing scheduler behaviour is risky for the existing dask user base.
2. Encoding annotations in the graph provides hints to upstream schedulers that they may respect, but are under no obligation to do so.
3. As it currently stands, users may
   - Create graphs via the collections (Array, Dataframe, Bag) and trust the scheduler to schedule work optimally.
   - Optimally schedule work imperatively via the distributed Client interface.
4. Neither approach seems entirely satisfactory: Collections are great abstractions which are disappointing to discard for the optimal compute/data placement afforded by the Client interface.
5. Annotations therefore bridge the gap between the two paradigms by providing an interface or glue layer.
6. They also provide the ability to build future intelligence into internal scheduling functionality such as dask.order.order. For example:
    - @eriknw might have some ideas about how to modify `dask.order.order` if provided with an estimate of the number of bytes that a task uses. Think ``Task(inc, 1, annotations={'nbytes': int(2**10)})``
    - provide custom task priorities within collection code, as demonstrated by @madsbk. Think ``Task(inc, 1, annotations={'priority': 100})``
    - Mark a task as dynamic for the functionality in https://github.com/dask/distributed/issues/3811.  @madsbk Think ``Task(inc, 1, annotations={'dynamic': True})``
    - The NVIDIA contingent might be interested in scheduling tasks on specific GPUs and streams. Think ``Task(inc, 1, annotations={'device': 'GPU0', 'cuda_stream': 0}``

I'm less interested in implementing the type of things in (6) as I think think Annotations provide sufficient flexibility for others to extend scheduling and write their own Scheduling Plugins.

Going forward I would be comfortable with:

1. Completing the implementation of my listed goals.
2. Invited comment and review on this implementation.
3. Keeping this branch up to date with master.
4. Maintaining this PR and https://github.com/dask/distributed/pull/3796
    1.  As a base for other contributor's PR's (i.e. in the vein of (6)).
    2. Reviewing such PRs
    3. to the point where the community and maintainers would be happy to perform a final merge.

How would the community feel about that?

**Comment by mrocklin** [comment]
Some personal thoughts:

1. As above, be ready for this to take a while.  There are a few stakeholders and it may be hard to come to consensus.
2. Maybe we think about how best to make smooth iterative progress on this topic.  I'd love for NVIDIA folks to weigh in here.
3. I'm curious about performance impacts

**Comment by madsbk** [comment]
Looks great @sjperkins.

My only concern is performance. If we decide on introducing a `Task` class, we should be able to use it extensively without any performance hit.

But if we can implement this PR with (almost) no performance penalty, I think this is exactly the smooth progress we are looking for -- the introduction of an optional `Task` class that initially doesn't impose any restrictions to the user but enables Dask components to _communicate_ through annotations. 

With this foundation, we can continue the discussion of the semantic of the annotated information.

**Comment by sjperkins** [comment]
I agree with the sentiments about maintaining existing performance.

Regarding annotation semantics, I don't feel strongly about imposing constraints on them, except w.r.t. to existing task kwargs in the Client.submit interface. I believe they should be treated as names "reserved" for use by the core schedulers because they already use them.

I air quote "reserved" because they're reserved by convention only: Their reservation and semantics should be mentioned in the Task object documentation.

Using typing terminology these currently reserved annotation names would look like:

```python
{
    'allow_other_workers': Bool,
    'priority': Int,
    'workers': Union[Tuple[Str], List[Str], Str],
    'resources': Union[Tuple[Str], List[Str], Str],
    'retries': Int,
}
```

The following name is not currently used at all, but I think should be "reserved" for future use as I think it work well with the Array and Dataframe collections:

```python
{ 'nbytes': Int }
```

I'll spitball some other annotations that may be sensible for future use in future posts. Perhaps others can think of some suggestions here?

**Comment by sjperkins** [comment]
# scatter

Instruct the scheduler to scatter task results to workers:

```python
{ 'scatter': Union[Bool, Str, Tuple[Str], List[Str]] }

{'scatter': True }            # scatter to all workers
{'scatter': "__all__" }       # scatter to all workers
{'scatter': 'alice'}          # scatter to worker 'alice
{'scatter': ['alice', 'bob'] } # scatter to alice and bob
```

**Comment by sjperkins** [comment]
Just a quick update on this PR: I think the functionality is in a good state.

1. The test cases pass on this PR and all except one on distributed [PR](https://github.com/dask/distributed/pull/2180)
2. The local scheduling code handles both Task objects and Task tuples
3. The distributed scheduler handles both Task objects and Task tuples
4. The optimization code handles both Task objects and Task tuples

I therefore consider the bulk of the implementation complete. I would still like to:

1. benchmark
2. demonstrate some use cases with the distributed scheduler and SchedulerPlugins

**Comment by sjperkins** [comment]
@mrocklin and others.

Following on from the meeting discussion, I thought I'd try and summarise the various approaches  to supporting newer Task style objects:

1 Task Objects and Tuple Tasks with Conversion
------------------------------------------------------------------

Pros:
 

- Flexible collection scheduling via scheduler plugins
- dask Complexity + Maintenance stay the same as only need to deal with new style tasks
- Supports old task graphs in downstream projects.
- Blockwise + SubGraphCallable's use new style Task objects

Cons:

- Task Conversion is an expense

  - When performed in cull for e.g. increases cull runtime by 6X in benchmarks

2  Task Objects and Tuple Tasks, no Conversion
------------------------------------------------------------------

Pros:

- Flexible collection scheduling via scheduler plugins
- Supports old task graphs in downstream projects
- Blockwise + SubGraphCallable's use new style Task objects

Cons:

- dask Complexity + Maintenance increases due to need to support both style tasks throughout codebase (+downstream)

  - subs benchmarks are ~20% slower supporting both in the current branch

3 Only Task Objects, no Conversion
---------------------------------------------------

Pros:

- Flexible collection scheduling via scheduler plugins
- Entire dask + distributed code base converted to new style Task objects
- No runtime task conversion required, I might guess that benchmarks would retain their performance characteristics.
- dask Complexity + Maintenance stay the same as only need to deal with new style tasks

Cons:

- Will break low-level graph compatibility for downstream projects (dask 3.0?)

Currently, this PR is a combination of (1) and (2) as it is predicated on not breaking downstream projects.  This results in either the expense of Conversion, or the expense of maintaining a code base with both Task objects and Tuple tasks.

I was intrigued by your idea of converting the entire dask code base to Task objects. The only downside to this that I can see is breaking downstream projects that use Tuple Tasks. However, the new graph format (dask 3.0?) would be simpler and I'd be interested in going forward with this approach. Another thought: the task conversion process (Task.from_spec, Task.from_call) could simply be repurposed as an upgrade path.

To summarise: This PR is at a three-way cross-roads, described above and it would be useful to identify what is desirable and what can be discarded, before going forward.

I'd be interested in hearing your thoughts on the subject. Perhaps an initial review is worthwhile at this point?

/cc @o-smirnov, @landmanbester, @jskenyon @kkraus14 @madsbk

**Comment by sjperkins** [comment]
I'm aware of and following #5644 and #6438

**Comment by sjperkins** [comment]
Note that https://github.com/dask/distributed/pull/2180 is the distributed sister PR and I've configured the CI in both PR's to refer to the opposite PR.

**Comment by jsignell** [comment]
@sjperkins is this PR still useful as a proof of concept given that there are now task annotations?

**Comment by sjperkins** [comment]
@jsignell I don't think I'll take it further, but it might be interesting for historical reasons. This would depend on the sjperkins fork remaining. I don't object to closing the PR.
